### PR TITLE
Update SMTP Test Email Logic

### LINKED ISSUE #3813: Testing SMTP uses old logic
## Description of bug
Provide a clear and concise description of the bug.
https://github.com/dnnsoftware/Dnn.Platform/pull/3237
## Steps to reproduce
List the precise steps to reproduce the bug:
1. send test emails with different configurations of portal/host
2. if host email is not set even when using portal smtp settings the test will fail

## Current behavior
The smtp test is using host settings to and from email addresses only for testing emails
Setting another domain that is not allowed to send emails as host email will potentially break portal smtp settings.

## Expected behavior
Portal settings for to and from addresses send emails from site security administrator account or host account.

## Error information
Blank Host Email in Host SMTP Settings will cause an error currently while trying to test smtp using portal settings.  Logged in as host during this test.  I set different email addresses

## Additional context

 I set different email addresses in a variety of ways and configurations to see which emails are being used and when and it reminded me of previous fix I had done.  I may have only tested while logged in as an administrator  and not as host so this is why it still acts this way while logged in as host.  It checks if host or not so things should work as intended similar to when you are logged in as admin (still need to test that theory of admin or host uses two different test SMTP methods/files as the names of the files may insist on) with the PR fix submitted. (still need to test that theory of admin or host uses two different test SMTP methods/files as the names of the files may insist on) 

This could be when logged in as host, when you are an admin maybe it worked.  I think the same logic should work in this location as well to set things right.  I know it worked before I will test as an administrator rather than host see if it makes a difference.

## Affected version

* [ ] 10.00.00 alpha build (in development)
* [ x ] 09.06.01 latest supported release

## Affected browser

* [ x ] Chrome
* [  x ] Firefox
* [  x] Safari
* [ x ] Internet Explorer 11
* [x ] Microsoft Edge (Classic)
* [ x] Microsoft Edge Chromium

### PR DESCRIPTION
**Pull request by thabaum**
Fixes #3813 

## Summary

Simplifies and makes the logic for TEST SMTP emails used **FROM** and **TO** and utilizes newer logic for both a Host or an Administrator when sending a test email.

Somehow the previous didnt get merged.  I think this code change refines things a bit to get the job done as expected.

### CODE REVIEW

**Review by bdukes — APPROVED** [review]

**Comment by bdukes** [comment]
@thabaum I think you need to leave that `smtpHostMode` variable, otherwise it looks good
```
Services\ServerSettingsSmtpAdminController.cs(118,21): error CS0103: The name 'smtpHostMode' does not exist in the current context [D:\a\1\s\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj]
Services\ServerSettingsSmtpAdminController.cs(119,21): error CS0103: The name 'smtpHostMode' does not exist in the current context [D:\a\1\s\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj]
Services\ServerSettingsSmtpAdminController.cs(120,21): error CS0103: The name 'smtpHostMode' does not exist in the current context [D:\a\1\s\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj]
Services\ServerSettingsSmtpAdminController.cs(121,21): error CS0103: The name 'smtpHostMode' does not exist in the current context [D:\a\1\s\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj]
Services\ServerSettingsSmtpAdminController.cs(122,21): error CS0103: The name 'smtpHostMode' does not exist in the current context [D:\a\1\s\Dnn.AdminExperience\Dnn.PersonaBar.Extensions\Dnn.PersonaBar.Extensions.csproj]
```

**Comment by thabaum** [comment]
```
                var smtpHostMode = request.SmtpServerMode == "h";
                var mailFrom = smtpHostMode ? Host.HostEmail : this.PortalSettings.Email;
                var mailTo = this.UserInfo.Email;
```

I believe I remember something with needing to set the variable `smtpHostMode` first then mail from so I will set them like they where.

**Comment by thabaum** [comment]
I believe this is what we had to do in the past so just making sure they both are applied to both files now.

**Comment by thabaum** [comment]
OK I took another look at those errors, The Host controller can probably work with the logic I had first had am I correct?  

And to use this same logic in the Admin ? I see what I am doing wrong here.  

Should we add 

```
                    smtpHostMode ? HostController.Instance.GetString("SMTPServer") : request.SmtpServer,
                    smtpHostMode ? HostController.Instance.GetString("SMTPAuthentication") : request.SmtpAuthentication.ToString(),
                    smtpHostMode ? HostController.Instance.GetString("SMTPUsername") : request.SmtpUsername,
                    smtpHostMode ? HostController.Instance.GetEncryptedString("SMTPPassword", Config.GetDecryptionkey()) : request.SmtpPassword,
                    smtpHostMode ? HostController.Instance.GetBoolean("SMTPEnableSSL", false) : request.EnableSmtpSsl);
```  
in place of 

logic in the HOST controller for same settings as well then am I correct?

File: Dnn.AdminExperience/Dnn.PersonaBar.Extensions/Services/ServerSettingsSmtpHostController.cs [hunk 1]
```diff
@@ -1,190 +1,191 @@
...
+                    Localization.GetSystemMessage(this.PortalSettings, "EMAIL_SMTP_TEST_SUBJECT"),
+                    MailFormat.Text,
+                    Encoding.UTF8,
+                    "",
+                    "",
+                    smtpHostMode ? HostController.Instance.GetString("SMTPServer") : request.SmtpServer,
+                    smtpHostMode ? HostController.Instance.GetString("SMTPAuthentication") : request.SmtpAuthentication.ToString(),
+                    smtpHostMode ? HostController.Instance.GetString("SMTPUsername") : request.SmtpUsername,
+                    smtpHostMode ? HostController.Instance.GetEncryptedString("SMTPPassword", Config.GetDecryptionkey()) : request.SmtpPassword,
+                    smtpHostMode ? HostController.Instance.GetBoolean("SMTPEnableSSL", false) : request.EnableSmtpSsl);
```
Comment (bdukes) [inline Dnn.AdminExperience/Dnn.PersonaBar.Extensions/Services/ServerSettingsSmtpHostController.cs:148]: I would think that we always want the test email to use the visible settings, regardless of mode.

**Comment by thabaum** [comment]
@bdukes I agree and that was a question I was pondering.  Remove smtpHostMode from both and use the code I originally submitted should resolve the smtpHostMode error as well.

I was for getting rid of the "mode" switch.  I added it to both rather than get rid of it but not sure it is doing anything different.

I can change the code to the following: 

```
                var mailFrom = request.SmtpServerMode == "h" ? Host.HostEmail : this.PortalSettings.Email;
                var mailTo = this.UserInfo.Email;

                var errMessage = Mail.SendMail(mailFrom,
                    mailTo,
                    "",
                    "",
                    MailPriority.Normal,
                    Localization.GetSystemMessage(this.PortalSettings, "EMAIL_SMTP_TEST_SUBJECT"),
                    MailFormat.Text,
                    Encoding.UTF8,
                    "",
                    "",
                    request.SmtpServer,
                    request.SmtpAuthentication.ToString(),
                    request.SmtpUsername,
                    request.SmtpPassword,
                    request.EnableSmtpSsl);

```

The `request.SmtpServer` should bring up the current setting for the portal being tested am I correct?  So why have the smtpHostMode variable in any of the code is the question I have if that is the case?

Should this be the best solution above then?

**Comment by bdukes** [comment]
It makes sense to me that in both cases it would be testing with the SMTP info entered on the page.

**Comment by thabaum** [comment]
OK So for an "Admin" they will not see the host admin settings so it must get them if it is in host mode.

For a Host they will see the settings for Host and Portal so they do not need to have the `smtpHostMode` set am I correct?

The admin using  the other code with `smtpHostMode` since they wont see the settings?

**Comment by thabaum** [comment]
Looking at this more closely now I have made the change to the Host file to not look for mode, however for the admin I think this is necessary as it wont be visible.

**Comment by thabaum** [comment]
I just cleaned up the files a little house keeping last changes:

Admin 

```
                var smtpHostMode = request.SmtpServerMode == "h";
                var mailFrom = smtpHostMode ? Host.HostEmail : this.PortalSettings.Email;
                var mailTo = this.UserInfo.Email;

                var errMessage = Mail.SendMail(mailFrom,
                    mailTo,
                    "",
                    "",
                    MailPriority.Normal,
                    Localization.GetSystemMessage(this.PortalSettings, "EMAIL_SMTP_TEST_SUBJECT"),
                    MailFormat.Text,
                    Encoding.UTF8,
                    "",
                    "",
                    smtpHostMode ? HostController.Instance.GetString("SMTPServer") : request.SmtpServer,
                    smtpHostMode ? HostController.Instance.GetString("SMTPAuthentication") : request.SmtpAuthentication.ToString(),
                    smtpHostMode ? HostController.Instance.GetString("SMTPUsername") : request.SmtpUsername,
                    smtpHostMode ? HostController.Instance.GetEncryptedString("SMTPPassword", Config.GetDecryptionkey()) : request.SmtpPassword,
                    smtpHostMode ? HostController.Instance.GetBoolean("SMTPEnableSSL", false) : request.EnableSmtpSsl);

```

Host

```
                var mailFrom = request.SmtpServerMode == "h" ? Host.HostEmail : this.PortalSettings.Email;
                var mailTo = this.UserInfo.Email;

                var errMessage = Mail.SendMail(mailFrom,
                    mailTo,
                    "",
                    "",
                    MailPriority.Normal,
                    Localization.GetSystemMessage(this.PortalSettings, "EMAIL_SMTP_TEST_SUBJECT"),
                    MailFormat.Text,
                    Encoding.UTF8,
                    "",
                    "",
                    request.SmtpServer,
                    request.SmtpAuthentication.ToString(),
                    request.SmtpUsername,
                    request.SmtpPassword,
                    request.EnableSmtpSsl);
```

Main thing was to use the right email for sending and to send it to the user testing.

Issue with the host is it would only send from the host email, although it should be sending from the Administrator email when set to portal SMTP settings, if not a mail server that can only send messages from addresses with the same domain (Not a relay) will fail.

File: Dnn.AdminExperience/Dnn.PersonaBar.Extensions/Services/ServerSettingsSmtpHostController.cs [hunk 2]
```diff
@@ -1,190 +1,191 @@
...
+                {
+                    success,
+                    errMessage,
+                    confirmationMessage =
+                        string.Format(
+                            Localization.GetString(
+                                "EmailSentMessage",
+                                Components.Constants.ServersResourcersPath),
+                            mailFrom,
+                            mailTo),
```
Comment (bdukes) [inline Dnn.AdminExperience/Dnn.PersonaBar.Extensions/Services/ServerSettingsSmtpHostController.cs:157]: @thabaum did you intend to remove the `errorMessageSendingTestEmail` message?
```suggestion
                    confirmationMessage =
                        success
                            ? string.Format(
                                Localization.GetString(
                                    "EmailSentMessage",
                                    Components.Constants.ServersResourcersPath),
                                mailFrom,
                                mailTo)
                            : Localization.GetString("errorMessageSendingTestEmail"),
```

**Comment by thabaum** [comment]
Thanks @bdukes catching that.  No I did not mean to change that.

**Review by bdukes — APPROVED** [review]

**Review by mitchelsellers — APPROVED** [review]
Great change!

**Pull request merged**